File: src/com/example/main/datastructure/java/queue/ArrayQueue.java
package com.example.main.datastructure.java.queue;

import com.example.main.datastructure.java.array.Array;

/**
 * 数组队列
 * Created by ls on 18/7/16.
 */
public class ArrayQueue<E> implements Queue<E> {
    private Array<E> array;

    public ArrayQueue() {
        array = new Array<>();
    }

    public ArrayQueue(int capacity) {
        array = new Array<>(capacity);
    }

    @Override
    public int getSize() {
        return array.getSize();
    }

    @Override
    public boolean isEmpty() {
        return array.isEmpty();
    }

    @Override
    public void enqueue(E e) {
        array.addLast(e);
    }

    public int getCapacity() {
        return array.getCapacity();
    }

    @Override
    public E dequeue() {
        return array.removeFirst();
    }

    @Override
    public E getFront() {
        return array.getFirst();
    }

    @Override
    public String toString() {
        StringBuilder builder = new StringBuilder();
        builder.append("Queue:");
        builder.append("front:[");


        int size = array.getSize();
        for (int i = 0; i < size; i++) {
            builder.append(array.get(i));
            if (i != size - 1) {
                builder.append(", ");
            }
        }
        builder.append("]tail");
        return builder.toString();
    }

    public static void main(String[] args) {
        ArrayQueue<Integer> arrayQueue = new ArrayQueue<>();
        for (int i = 0; i < 10; i++) {
            arrayQueue.enqueue(i);
            System.out.println(arrayQueue);
            if (i % 3 == 0) {
                System.out.println(arrayQueue.dequeue());
            }
            System.out.println(arrayQueue);
        }
    }
}

File: src/com/example/main/book/Queue/ArrayQueue.kt
package com.example.main.book.Queue

import java.lang.Exception

class ArrayQueue {
    private var front: Int
    private var rear: Int
    private var capacity: Int
    private var array: Array<Int>

    private constructor(size: Int) {
        capacity = size
        front = -1
        rear = -1
        array = Array(capacity) { 0 }
    }

    fun isEmpty(): Boolean = front == -1

    fun isFull(): Boolean = (rear + 1) % capacity == front

    fun getQueueSize(): Int = (capacity - front + rear + 1) % capacity

    fun enQueue(data: Int) {
        if (isFull()) {
            throw Exception("Queue Overflow")
        } else {
            rear = (rear + 1) % capacity
            array[rear] = data
            if (front == -1) {
                front = rear
            }
        }
    }

    fun deQueue():Int{
        if (isEmpty()){
            throw Exception("Queue Empty")
        }else{
            val data = array[front]
            front = if (front == rear){
                rear-1
            }else{
                (front+1)%capacity
            }
            return data
        }
    }

    companion object {
        fun createQueue(size: Int): ArrayQueue {
            return ArrayQueue(size)
        }
    }
}

fun main(args: Array<String>) {
    val createQueue = ArrayQueue.createQueue(8)
    (0..7).forEach { createQueue.enQueue(it) }
    (0..7).forEach { _ -> println(createQueue.deQueue()) }
}